File: 프로그래머스/1/42862. 체육복/체육복.py
def solution(n, lost, reserve):
    # 여벌이 있지만 도난당한 학생을 먼저 제외 (차집합 사용)
    lost_set = set(lost) - set(reserve)
    reserve_set = set(reserve) - set(lost)
    
    # 체육복을 빌릴 수 있는 경우 찾기
    for miss in sorted(lost_set):
        if miss - 1 in reserve_set:  # 앞 번호 학생이 여벌 있음
            reserve_set.remove(miss - 1)
            lost_set.remove(miss)
        elif miss + 1 in reserve_set:  # 뒷 번호 학생이 여벌 있음
            reserve_set.remove(miss + 1)
            lost_set.remove(miss)
    
    return n - len(lost_set)  # 도난당한 학생 중 체육복을 못 빌린 학생 제외

File: 프로그래머스/1/42862. 체육복/체육복.java
import java.util.*;
class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        boolean[] isHaveCloth = new boolean[n];
        boolean[] isRemain = new boolean[n];
        Arrays.fill(isHaveCloth, true);
        Arrays.sort(lost);
        Arrays.sort(reserve);
        for (int r : reserve){
            isRemain[r-1] = true;
        }
        
        for (int l : lost){
            isHaveCloth[l-1] = false;
        }
        
        for (int l : lost){
            if (isRemain[l - 1]) {
                isHaveCloth[l - 1] = true; 
                isRemain[l - 1] = false;  
            }
        }
        
        for (int l : lost){
            if (isHaveCloth[l-1]){
                continue;
            }
            
            // 앞번호
            if (l-2>=0 && isRemain[l-2]){
                isRemain[l-2] = false;
                isHaveCloth[l-1] = true;
            } 
            // 뒷번호
            else if (l < n && isRemain[l]){
                isRemain[l] = false;      
                isHaveCloth[l-1] = true;
            }
        }
        
        int answer = 0;
        for (boolean bool : isHaveCloth) {
            if (bool) answer++;
        }
        
        return answer;
    }
}